overview/cli.md

```diff
@@ -0,0 +1,369 @@
+
+There are two main topics when building a CLI application:
+* [input](#input)
+* [output](#output)
+
+## Input
+This topic covers all things related to:
+* [options passed to the app](#options)
+* [request for user input](#request-for-user-input)
+
+### Options
+It is a very common practice to pass options to the application. For example, we may run `crystal -v` and Crystal will display:
+
+```shell-session
+$ crystal -v
+Crystal 0.31.1 (2019-10-02)
+
+LLVM: 8.0.1
+Default target: x86_64-apple-macosx
+```
+
+and if we run: `crystal -h`, then Crystal will show all the accepted options and how to use them.
+
+So now the question would be: **do we need to implement an options parser?** No need to, Crystal got us covered with the class `OptionParser`. Let’s build an application using this parser!
+
+To begin with our CLI application will have two options:
+* `-v` / `--version`: it will display the application version.
+* `-h` / `--help`: it will display the application help.
+
+```crystal
+# file: all_my_cli.cr
+require "option_parser"
+
+OptionParser.parse do |parser|
+  parser.banner = "Welcome to The Beatles App!"
+
+  parser.on "-v", "--version", "Show version" do
+    puts "version 1.0"
+    exit
+  end
+  parser.on "-h", "--help", "Show help" do
+    puts parser
+    exit
+  end
+end
+```
+
+So, how does all this work? Well … magic! No, it’s not really magic! Just Crystal making our life easy.
+When our application starts, the block passed to `OptionParser#parse` gets executed. In that block we define all the options. After the block is executed, the parser will start consuming the arguments passed to the application, trying to match each one with the options defined by us. If an option matches then the block passed to `parser#on` gets executed!
+
+We can read all about `OptionParser` in [the official API documentation](https://crystal-lang.org/api/latest/OptionParser.html). And from there we are one click away from the source code ... the actual proof that it is not magic!
+
+Now, let's run our application. We have two ways [using the compiler](https://crystal-lang.org/reference/using_the_compiler):
+1. [Build the application](https://crystal-lang.org/reference/using_the_compiler/#crystal-build) and then run it.
+2. Compile and [run the application](https://crystal-lang.org/reference/using_the_compiler/#crystal-run), all in one command.
+
+We are going to use the second way:
+
+```shell-session
+$ crystal ./all_my_cli.cr -- -h
+
+Welcome to The Beatles App!
+    -v, --version                    Show version
+    -h, --help                       Show help
+```
+
+Let's continue building our _fabulous_ application adding a new feature:
+by default (i.e. no options given) the application will display the names of the Fab Four. But, if we pass the option `-t` / `--twist` it will display the names in uppercase:
```

Comment:
```suggestion
By default (i.e. no options given) the application will display the names of the Fab Four. But, if we pass the option `-t` / `--twist` it will display the names in uppercase:
```